library/pkcs7.c

```
@@ -0,0 +1,525 @@
+        return( MBEDTLS_ERR_PKCS7_INVALID_VERSION + ret );
+
+    return( 0 );
+}
+
+/**
+ * ContentInfo ::= SEQUENCE {
+ *      contentType ContentType,
+ *      content
+ *              [0] EXPLICIT ANY DEFINED BY contentType OPTIONAL }
+ **/
+static int pkcs7_get_content_info_type( unsigned char **p, unsigned char *end,
+                                        mbedtls_pkcs7_buf *pkcs7 )
+{
+    size_t len = 0;
+    int ret;
+
+    ret = mbedtls_asn1_get_tag( p, end, &len, MBEDTLS_ASN1_CONSTRUCTED
+                                            | MBEDTLS_ASN1_SEQUENCE );
+    if( ret )
+        return( MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO + ret );
+
+    ret = mbedtls_asn1_get_tag( p, end, &len, MBEDTLS_ASN1_OID );
+    if( ret )
+        return( MBEDTLS_ERR_PKCS7_INVALID_CONTENT_INFO + ret );
+
+    pkcs7->tag = MBEDTLS_ASN1_OID;
+    pkcs7->len = len;
+    pkcs7->p = *p;
+
+    return( ret );
+}
+
+/**
+ * DigestAlgorithmIdentifier ::= AlgorithmIdentifier
+ *
+ * This is from x509.h
+ **/
+static int pkcs7_get_digest_algorithm( unsigned char **p, unsigned char *end,
+                                       mbedtls_x509_buf *alg )
+{
+    int ret;
+
+    if( ( ret = mbedtls_asn1_get_alg_null( p, end, alg ) ) != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_ALG + ret );
+
+    return( 0 );
+}
+
+/**
+ * DigestAlgorithmIdentifiers :: SET of DigestAlgorithmIdentifier
+ **/
+static int pkcs7_get_digest_algorithm_set( unsigned char **p,
+                                           unsigned char *end,
+                                           mbedtls_x509_buf *alg )
+{
+    size_t len = 0;
+    int ret;
+
+    ret = mbedtls_asn1_get_tag( p, end, &len, MBEDTLS_ASN1_CONSTRUCTED
+                                            | MBEDTLS_ASN1_SET );
+    if( ret != 0 )
+        return( MBEDTLS_ERR_PKCS7_INVALID_ALG + ret );
+
+    end = *p + len;
+
+    /** For now, it assumes there is only one digest algorithm specified **/
```

The current code silently ignores the presence of further algorithm identifiers. Is this intended, or do you want to signal an error in this case?

---

> That said, maybe we should return a specific error code for this assumption - a generic `MBEDTLS_ERR_PKCS7_INVALID_FORMAT` might confuse users who supply a valid PKCS7 file. Something like `MBEDTLS_ERR_PKCS7_UNSUPPORTED_FORMAT` might be clearer.

Using `MBEDTLS_ERR_PKCS7_FEATURE_UNAVAILABLE` as it is already defined and seems to fit the bill.